### PR TITLE
ci: Extract and reuse existing workflow steps

### LINKED ISSUE #6091: Make use of CI build matrix in reused workflows
With #5764, lots of duplicated code was reduced, but more can be done there.
The docker build steps are still duplicated in CI, pre-release and release pipelines. This could be simplified by re-using the build matrix pattern from the CI pipelines and passing over the build matrix to a re-used workflow that contains the steps to build all docker images.
The new docker build workflow could then also be re-used in the pre-release and release pipeline.

Some references:
https://evilmartians.com/chronicles/build-images-on-github-actions-with-docker-layer-caching
https://docs.github.com/en/actions/learn-github-actions/reusing-workflows
https://github.community/t/how-to-share-matrix-between-jobs/128595/5

### PR DESCRIPTION
**Pull request by philipp-hinteregger**
Signed-off-by: Philipp Hinteregger <philipp.hinteregger@dynatrace.com>

## This PR

- extracts Unit Test and Docker build and push steps into a separate workflow
- reuses the new workflow in the CI / pre-release / release pipelines

### Related Issues

Fixes #6091

### CODE REVIEW

**Review by bacherfl — DISMISSED** [review]

File: .github/workflows/test-and-build-docker-images.yml [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    - name: Bridge lint checks
+      if: inputs.artifact == 'bridge2' && inputs.docker-test-target == 'builder-code-style'
+      working-directory: bridge
+      run: yarn prettier:check && yarn lint:check
+
+    - name: UI-Test
+      if: inputs.docker-test-target == 'builder-test-ui'
+      working-directory: bridge
+      run: |
+       ./cypress/run-tests.sh
```
Comment (mowies) [inline .github/workflows/test-and-build-docker-images.yml:94]:
```suggestion
       ./cypress/run-tests.sh
```

File: .github/workflows/CI.yml [hunk 2]
```diff
@@ -234,134 +241,32 @@ jobs:
...
+    secrets: inherit
+    with:
+      artifact: ${{ matrix.config.artifact }}
+      tags: |
+        keptndev/${{ matrix.config.artifact }}:${{ needs.prepare_ci_run.outputs.VERSION }}
+        keptndev/${{ matrix.config.artifact }}:${{ needs.prepare_ci_run.outputs.VERSION }}.${{ needs.prepare_ci_run.outputs.DATE }}${{ needs.prepare_ci_run.outputs.TIME }}
+      should-run: ${{ matrix.config.should-run }}
+      should-push-image: ${{ matrix.config.should-push-image == 'true' && (( github.event_name == 'pull_request' && github.event.pull_request.head.repo.full_name == github.repository ) || ( github.event_name == 'push' )) && github.actor != 'renovate[bot]' && github.actor != 'dependabot[bot]' }}
+      docker-test-target: ${{ matrix.config.docker-test-target }}
+      working-dir: ${{ matrix.config.working-dir }}
```
Comment (mowies) [inline .github/workflows/CI.yml:263]: why are you passing the matrix values through separately when you also define the matrix above as well?
is it not enough to just define the matrix?

File: .github/workflows/CI.yml [hunk 3]
```diff
@@ -200,11 +200,18 @@ jobs:
...
+          GO_VERSION: ${{ env.GO_VERSION }}
         run: |
           echo "::set-output name=DATE::$(date +'%Y%m%d')"
           echo "::set-output name=TIME::$(date +'%H%M')"
           echo "::set-output name=DATETIME::$(date +'%Y%m%d')$(date +'%H%M')"
 
+      - name: Get GO-Version
+        id: get_go_version
+        run: |
+          echo "::set-output name=GO_VERSION::$GO_VERSION"
```
Comment (mowies) [inline .github/workflows/CI.yml:213]: I don't think you need this the go version is defined globally in the workflow at the top

File: .github/workflows/CI.yml (hunk 2 above)
Comment (philipp-hinteregger) [reply .github/workflows/CI.yml:263]: I used this approach for future reusability,  so you wouldn't need a matrix to run this workflow.
In this case the matrix strategy is applied on the invoking workflow stage.
The alternative would be to pass the whole matrix into the reusable workflow and apply the strategy there.

**Review by mowies — DISMISSED** [review]
LGTM! good job :)

**Review by mowies — APPROVED** [review]

**Pull request merged**